File: binaryTrees/39.py
from typing import Optional, List


class TreeNode:
    def __init__(self, k: int):
        self.val = k
        self.left: TreeNode = None
        self.right: TreeNode = None


def sumEvenGrandparent(root: Optional[TreeNode]) -> int:
    sum: List[int] = [0]

    def dfs(
        root: Optional[TreeNode],
        parent: Optional[TreeNode],
        grandpa: Optional[TreeNode],
    ):
        if root is None:
            return

        if grandpa is not None and grandpa.val % 2 == 0:
            sum[0] += root.val

        dfs(root.left, root, parent)
        dfs(root.right, root, parent)

    dfs(root, None, None)

    return sum[0]


if __name__ == "__main__":
    root = TreeNode(6)
    root.left = TreeNode(7)
    root.right = TreeNode(8)
    root.left.left = TreeNode(2)
    root.left.right = TreeNode(7)
    root.left.left.left = TreeNode(9)
    root.left.right.left = TreeNode(1)
    root.left.right.right = TreeNode(4)
    root.right.left = TreeNode(1)
    root.right.right = TreeNode(3)
    root.right.right.right = TreeNode(5)

    print(sumEvenGrandparent(root))

File: binaryTrees/39.cpp
#include "../common/tree.h"

using namespace Tree;

void dfs(TreeNode<int> *root, int &sum, TreeNode<int> *parent, TreeNode<int> *grandpa) {
    if (root == nullptr)
        return;

    if (grandpa != nullptr && (grandpa->val & 1) == 0)
        sum += root->val;

    dfs(root->left, sum, root, parent);
    dfs(root->right, sum, root, parent);
}

int sumEvenGrandparent(TreeNode<int> *root) {
    int sum = 0;
    dfs(root, sum, nullptr, nullptr);
    return sum;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    TreeNode<int> *root = createNode(6);
    root->left = createNode(7);
    root->right = createNode(8);
    root->left->left = createNode(2);
    root->left->right = createNode(7);
    root->left->left->left = createNode(9);
    root->left->right->left = createNode(1);
    root->left->right->right = createNode(4);
    root->right->left = createNode(1);
    root->right->right = createNode(3);
    root->right->right->right = createNode(5);

    cout<< sumEvenGrandparent(root);

    return 0;
}